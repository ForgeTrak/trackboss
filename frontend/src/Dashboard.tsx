import React from 'react';
import './App.css';
import { ChakraProvider } from '@chakra-ui/react';
import theme from './theme';
import Header from './components/Header';

function Dashboard() {
    return (
        <ChakraProvider theme={theme}>
            <Header title="Dashboard" />
        </ChakraProvider>
    );
}

export default Dashboard;
