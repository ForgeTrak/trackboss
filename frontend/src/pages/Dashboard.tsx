import React, { useEffect, useState, useContext } from 'react';
import { ChakraProvider, Center, SimpleGrid, VStack } from '@chakra-ui/react';
import { UserContext } from '../contexts/UserContext';
import theme from '../theme';
import Header from '../components/Header';
import WorkPointsCard from '../components/WorkPointsCard';
import ImportantLinksCard from '../components/ImportantLinksCard';
import EventCard from '../components/EventCard';
import { getEventCardProps } from '../controller/event';
import { getWorkPointsTotal } from '../controller/workPoints';
import { getYearlyThresholdValue } from '../controller/billing';
import GreetingText from '../components/GreetingText';
import { getTodaysDate } from '../controller/utils';
import { getGateCodeLatest } from '../controller/gateCode';
import { GateCode } from '../../../src/typedefs/gateCode';

async function getEventCardPropsLocal(token: string): Promise<any | undefined> {
    const nowString = getTodaysDate();
    const props = await getEventCardProps(token, `${nowString}-`);
    return props;
}

async function getWorkPointsPercentage(token: string, memberId: number) {
    const workPoints = await getWorkPointsTotal(token, memberId);
    const threshold = await getYearlyThresholdValue(token);
    if (workPoints && threshold) {
        return Math.ceil((workPoints / threshold) * 100);
    }
    // else
    return 0;
}

function Dashboard() {
    const { state } = useContext(UserContext);
    const [eventCardProps, setEventCardProps] = useState<any>();
    const [percent, setPercent] = useState<number>(0);
    const [gateCode, setGateCode] = useState<string>('');

    useEffect(() => {
        async function getData() {
            setEventCardProps(await getEventCardPropsLocal(state.token));
            if (state.user) {
                setPercent(await getWorkPointsPercentage(state.token, state.user.memberId));
            }
        }
        getData();
    }, [state.user]);

    useEffect(() => {
        async function getData() {
            const latestGateCode = await getGateCodeLatest(state.token) as GateCode;
            setGateCode(latestGateCode.gateCode);
        }
        getData();
    });

    return (
        <ChakraProvider theme={theme}>
            <VStack align="left" spacing="2em">
                <Header title="Track Boss Dashboard" activeButtonId={1} />
                {
                    state.user && (
                        <GreetingText name={`${state.user.firstName} ${state.user.lastName}`} />
                    )
                }
                <Center>
                    <SimpleGrid columns={[1, null, 3]} spacing="20px">
                        <WorkPointsCard percent={percent} />
                        {
                            eventCardProps ? (
                                <EventCard
                                    date={eventCardProps.start}
                                    startTime={eventCardProps.time}
                                    name={eventCardProps.title}
                                />
                            ) : (
                                <EventCard
                                    date=""
                                    startTime=""
                                    name=""
                                />
                            )
                        }
                        <ImportantLinksCard gateCode={gateCode} />
                    </SimpleGrid>
                </Center>
            </VStack>
        </ChakraProvider>
    );
}

export default Dashboard;
