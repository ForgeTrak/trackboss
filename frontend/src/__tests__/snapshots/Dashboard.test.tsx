import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Dashboard from '../../pages/Dashboard';

it('renders correctly', () => {
    const dash = renderer.create(
        <BrowserRouter>
            <Dashboard />
        </BrowserRouter>,
    ).toJSON();
    expect(dash).toMatchSnapshot();
});