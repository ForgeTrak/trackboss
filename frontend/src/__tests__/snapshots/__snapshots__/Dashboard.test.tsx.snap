// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dashboard renders correctly 1`] = `
Array [
  <div
    className="chakra-stack css-p747s7"
  >
    <div>
      <div
        className="css-arm2n8"
      >
        <div
          className="chakra-stack css-84zodg"
        >
          <button
            aria-label="Back"
            className="chakra-button css-76fq9n"
            onClick={[Function]}
            type="button"
          >
            <svg
              aria-hidden={true}
              fill="currentColor"
              focusable={false}
              height="lg"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="lg"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M217.9 256L345 129c9.4-9.4 9.4-24.6 0-33.9-9.4-9.4-24.6-9.3-34 0L167 239c-9.1 9.1-9.3 23.7-.7 33.1L310.9 417c4.7 4.7 10.9 7 17 7s12.3-2.3 17-7c9.4-9.4 9.4-24.6 0-33.9L217.9 256z"
              />
            </svg>
          </button>
          <div>
            <button
              aria-label="Menu"
              className="chakra-button css-1y85x9m"
              onClick={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                fill="currentColor"
                focusable={false}
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M904 160H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0 624H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8zm0-312H120c-4.4 0-8 3.6-8 8v64c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-64c0-4.4-3.6-8-8-8z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          className="css-17xejub"
        />
        <div
          className="css-0"
        >
          <h2
            className="chakra-heading css-1c7bccy"
          >
            Dashboard
          </h2>
        </div>
        <div
          className="css-17xejub"
        />
      </div>
    </div>
    <div
      className="css-gmuwbf"
    >
      <div
        className="chakra-stack css-84zodg"
      >
        <div
          className="css-15ddyrp"
          onClick={[Function]}
        >
          <div
            className="css-gmuwbf"
          >
            <div
              className="chakra-stack css-10j85wf"
            >
              <GaugeChart
                colors={
                  Array [
                    "#FF9F46",
                    "#76CE6F",
                  ]
                }
                hideText={true}
                id="gauge-chart"
                nrOfLevels={20}
                percent={0.67}
              />
              <div
                className="chakra-stack css-jp20jv"
              >
                <p
                  className="chakra-text css-1srjgni"
                >
                  You have completed
                </p>
                <p
                  className="chakra-text css-68kehq"
                >
                  0
                  %
                </p>
                <p
                  className="chakra-text css-1bsg0uq"
                >
                  of this year's work points
                </p>
              </div>
            </div>
          </div>
        </div>
        <div />
        <div
          className="css-1z0glir"
        >
          <div
            className="css-gmuwbf"
          >
            <div
              className="chakra-stack css-nscsg6"
            >
              <h2
                className="chakra-heading css-1dklj6k"
              >
                Important Links
              </h2>
              <hr
                aria-orientation="horizontal"
                className="chakra-divider css-svjswr"
              />
              <ul
                className="css-1avoc21"
                role="list"
              >
                <a
                  className="chakra-link css-k6vkpw"
                  href="https://palmyramx.com/pages/track-rules"
                  target="_blank"
                >
                  <li
                    className="css-81gbng"
                  >
                    Member Rules
                  </li>
                </a>
                <a
                  className="chakra-link css-k6vkpw"
                  href="https://twitter.com/palmyramx"
                  target="_blank"
                >
                  <li
                    className="css-81gbng"
                  >
                    Twitter
                  </li>
                </a>
                <a
                  className="chakra-link css-k6vkpw"
                  href="https://www.facebook.com/palmyramx/"
                  target="_blank"
                >
                  <li
                    className="css-81gbng"
                  >
                    Facebook
                  </li>
                </a>
                <a
                  className="chakra-link css-k6vkpw"
                  href="https://www.instagram.com/palmyramx_hogback_hill/?hl=en"
                  target="_blank"
                >
                  <li
                    className="css-81gbng"
                  >
                    Instagram
                  </li>
                </a>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <span />,
]
`;
